1. Remove the termination callback features from pybitwuzla. They depend on a
   private header file, utils/bzlaabort.h, that isn't available to standalone
   Cython builds.

2. Correct errors in the type annotations.

3. Remove use of array.array, which is incompatible with PyPy.

--- pybitwuzla.pyx	2023-07-06 20:43:01.308048822 +0000
+++ pybitwuzla.pyx	2023-07-06 20:43:10.067621654 +0000
@@ -18,9 +18,7 @@
 from libc.stdint cimport int32_t, uint32_t, uint64_t
 from libcpp cimport bool as cbool
 from cpython.ref cimport PyObject
-from cpython cimport array
 from collections import defaultdict
-import array
 import math, os, sys
 import tempfile

@@ -250,7 +248,7 @@
            Get string representation of term in format ``fmt``.

            :param fmt: Output format. Available formats: "btor", "smt2"
-           :type: str
+           :type fmt: str

            :return: String representation of the term in format ``fmt``.
            :rtype: str
@@ -297,7 +295,7 @@

     def get_symbol(self):
         """:return: The symbol of the term.
-           :rtype: str
+           :rtype: str or None

            .. seealso::
                :func:`~pybitwuzla.BitwuzlaTerm.set_symbol`
@@ -479,69 +477,12 @@

     def __dealloc__(self):
         if self._c_bitwuzla is not NULL:
-            bitwuzla_api.pybitwuzla_delete(self._c_bitwuzla)
+            bitwuzla_api.bitwuzla_delete(self._c_bitwuzla)

     cdef bitwuzla_api.Bitwuzla* ptr(self):
         return self._c_bitwuzla

     # ------------------------------------------------------------------------
-    # Termination callback
-    # ------------------------------------------------------------------------
-
-    def set_term(self, fun, args):
-        """set_term(fun, args)
-
-           Set a termination callback function.
-
-           Use this function to force Bitwuzla to prematurely terminate if
-           callback function ``fun`` returns True. Arguments ``args`` to
-           ``fun`` may be passed as a single Python object (in case that
-           ``fun`` takes only one argument), a tuple, or a list of arguments.
-
-           E.g., ::
-
-             import time
-
-             def fun1 (arg):
-                 # timeout after 1 sec.
-                 return time.time() - arg > 1.0
-
-             def fun2 (arg0, arg1):
-                 # do something and return True/False
-                 ...
-
-             bitwuzla = Bitwuzla()
-
-             bitwuzla.set_term(fun1, time.time())
-             bitwuzla.set_term(fun1, (time.time(),))
-             bitwuzla.set_term(fun1, [time.time()])
-
-             bitwuzla.set_term(fun2, (arg0, arg1))
-             bitwuzla.set_term(run2, [arg0, arg1])
-
-           :param fun: A python function.
-           :param args: A function argument or a list or tuple of function
-                        arguments.
-        """
-        cdef PyObject* funptr = <PyObject*>fun
-        cdef PyObject* argsptr = <PyObject*>args
-        bitwuzla_api.pybitwuzla_set_term(self.ptr(), funptr, argsptr)
-
-    def terminate(self):
-        """Call terminate callback that was set via
-           :func:`~pybitwuzla.Bitwuzla.set_term`.
-
-           :return: True if termination condition is fulfilled, else False.
-           :rtype: bool
-
-           .. seealso::
-                :func:`~pybitwuzla.Bitwuzla.set_term`.
-        """
-        cdef int32_t res
-        res = bitwuzla_api.bitwuzla_terminate(self.ptr())
-        return res > 0
-
-    # ------------------------------------------------------------------------
     # Bitwuzla API functions (general)
     # ------------------------------------------------------------------------

@@ -778,7 +719,7 @@
            Get the model as a string in format ``fmt``.

            :param fmt: Model format. Available formats: "btor", "smt2"
-           :type: str
+           :type fmt: str

            :return: String representation of model in format ``fmt``.
            :rtype: str
@@ -797,7 +738,7 @@
            Dump the current formula as a string in format ``fmt``.

            :param fmt: Model format. Available formats: "btor", "smt2"
-           :type: str
+           :type fmt: str

            :return: String representation of formula in format ``fmt``.
            :rtype: str
@@ -917,6 +858,7 @@
            :param opt:   Option.
            :type opt:    BitwuzlaOption
            :param value: Option value.
+           :type value:  str or BitwuzlaOption

            .. seealso::
                 For a list of available options see :class:`~pybitwuzla.Option`
@@ -1388,17 +1330,15 @@
                                                      sort.ptr(),
                                                      _to_cstr(symbol)))

-    def mk_term(self, kind, terms, indices = None):
-        """mk_term(kind, terms, indices = None)
+    def mk_term(self, kind, terms):
+        """mk_term(kind, terms)

            Create a term of given kind with the given argument terms.

            :param kind: The operator kind.
            :type kind: Kind
            :param terms: The number of argument terms.
-           :type terms: list(BitwuzlaTerm)
-           :param indices: The argument terms.
-           :type indices: list(int)
+           :type terms: list(BitwuzlaTerm) or tuple(BitwuzlaTerm)

            :return: A term representing an operation of given kind.
            :rtype: BitwuzlaTerm
@@ -1407,10 +1347,6 @@
             raise ValueError('Given kind is not a Kind object.')
         if not isinstance(terms, list) and not isinstance(terms, tuple):
             raise ValueError('Expected list or tuple for terms')
-        if indices is not None \
-            and not isinstance(indices, list) \
-            and not isinstance(indices, tuple):
-            raise ValueError('Expected list or tuple for indices')

         num_terms = len(terms)
         cdef const bitwuzla_api.BitwuzlaTerm **c_terms =\
@@ -1423,20 +1359,8 @@
             c_terms[i] = (<BitwuzlaTerm> terms[i]).ptr()

         term = BitwuzlaTerm(self)
-
-        cdef array.array c_indices
-        if indices:
-            c_indices = array.array('I', indices)
-            term.set(bitwuzla_api.bitwuzla_mk_term_indexed(
-                            self.ptr(),
-                            kind.value,
-                            num_terms,
-                            c_terms,
-                            len(indices),
-                            c_indices.data.as_uints))
-        else:
-            term.set(bitwuzla_api.bitwuzla_mk_term(
-                        self.ptr(), kind.value, num_terms, c_terms))
+        term.set(bitwuzla_api.bitwuzla_mk_term(
+                    self.ptr(), kind.value, num_terms, c_terms))
         free(c_terms)
         return term

@@ -1448,7 +1372,7 @@
            substitutions in ``subst_map``.

            :param terms: List of terms to apply substitutions.
-           :type terms: list(BitwuzlaTerm)
+           :type terms: list(BitwuzlaTerm) or tuple(BitwuzlaTerm) or BitwuzlaTerm
            :param subst_map: The substitution map mapping constants or
                              variables to terms.
            :type subst_map: dict(BitwuzlaTerm,BitwuzlaTerm)
